alibi_detect/utils/loading.py

```diff
@@ -0,0 +1,625 @@
+
+
+def _load_optimizer(cfg: dict,
+                    backend: str) -> Union['tf.keras.optimizers.Optimizer', Callable]:
+    """
+    Loads an optimzier from an optimizer config dict. When backend='tensorflow', the config dict should be in
+    the format given by tf.keras.optimizers.serialize().
+
+    Parameters
+    ----------
+    cfg
+        The optimizer config dict.
+    backend
+        The backend.
+
+    Returns
+    -------
+    The loaded optimizer.
+    """
+    if backend == 'tensorflow':
+        optimizer = load_optimizer_tf(cfg)
+    else:
+        raise NotImplementedError('Loading of pytorch optimizers not currently supported')
+    return optimizer
+
+
+def _get_nested_value(dic: dict, keys: list) -> Any:
+    """
+    Get a value from a nested dictionary.
+
+    Parameters
+    ----------
+    dic
+        The dictionary.
+    keys
+        List of keys to "walk" to nested value.
+        For example, to extract the value `dic['key1']['key2']['key3']`, set `keys = ['key1', 'key2', 'key3']`.
+
+    Returns
+    -------
+    The nested value specified by `keys`.
+    """
+    for key in keys:
+        try:
+            dic = dic[key]
+        except (TypeError, KeyError):
+            return None
+    return dic
+
+
+def _set_nested_value(dic: dict, keys: list, value: Any):
+    """
+    Set a value in a nested dictionary.
+
+    Parameters
+    ----------
+    dic
+        The dictionary.
+    keys
+        List of keys to "walk" to nested value.
+        For example, to set the value `dic['key1']['key2']['key3']`, set `keys = ['key1', 'key2', 'key3']`.
+    value
+        The value to set.
+    """
+    for key in keys[:-1]:
+        dic = dic.setdefault(key, {})
+    dic[keys[-1]] = value
+
+
+def read_config(filepath: Union[os.PathLike, str]) -> dict:
+    """
+    This function reads a detector toml config file and returns a dict specifying the detector.
+
+    Parameters
+    ----------
+    filepath
+        The filepath to the config.toml file.
+
+    Returns
+    -------
+        Parsed toml dictionary.
+    """
+    filepath = Path(filepath)
+
+    cfg = dict(toml.load(filepath))  # toml.load types return as MutableMapping, force to dict
+    logger.info('Loaded config file from %s', str(filepath))
+
+    # Convert keys in categories_per_feature back to str. This is undesirable, but currently necessary in order
+    # to use toml library, as error with toml dumping dict with int keys.
+    if 'categories_per_feature' in cfg:
+        new = {}
+        for key in cfg['categories_per_feature']:
+            new[int(key)] = cfg['categories_per_feature'][key]
+        cfg['categories_per_feature'] = new
+
+    # This is necessary as no None/null in toml spec., and missing values are set to defaults set in pydantic models.
+    # But we sometimes to do explicitly spec as None.
+    cfg = _replace(cfg, "None", None)
+
+    return cfg
+
+
+def resolve_cfg(cfg: dict, config_dir: Optional[Path]) -> dict:
```

Comment:
Let's try to keep consistent with names, e.g. currently `read_config` vs `resolve_cfg`.